### PR TITLE
Validate locale and external file with data annotation.

### PR DESCRIPTION
**Pull request by live1206**

### CODE REVIEW

File: src/docfx/Errors.cs [hunk 1]
```diff
@@ -19,12 +19,6 @@ public static Error InvalidRedirection(string path, ContentType contentType)
         public static Error ConfigNotFound(string docsetPath)
             => new Error(ErrorLevel.Error, "config-not-found", $"Cannot find docfx.yml at '{docsetPath}'");
```
Comment (yufeih) [inline src/docfx/Errors.cs]: also remove `GitCommitsTimeNotFound` and `UserProfileCacheNotFound`

File: src/docfx/Errors.cs [hunk 2]
```diff
@@ -127,6 +121,9 @@ public static Error UnknownField(Range range, string propName, string typeName,
         public static Error ViolateSchema(Range range, string message)
             => new Error(ErrorLevel.Error, "violate-schema", $"{range} {message}", line: range.StartLine, column: range.StartCharacter);
 
+        public static Error ValidationContextMissing(string name)
```
Comment (yufeih) [inline src/docfx/Errors.cs]: remove this error-code

File: src/docfx/lib/JsonUtility.cs [hunk 3]
```diff
@@ -98,22 +100,24 @@ public static (List<Error>, T) Deserialize<T>(string json)
         /// Creates an instance of the specified .NET type from the JToken
         /// And validate mismatching field types
         /// </summary>
-        public static (List<Error>, T) ToObject<T>(JToken token)
+        public static (List<Error>, T) ToObject<T>(JToken token, Dictionary<object, object> items = null)
```
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: consider passing an optional `docsetPath` directly

File: src/docfx/lib/attributes/ExternalFileAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
```
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/lib/attributes/ExternalFileAttribute.cs]: The reason I remain them under docfx is that there is a dependency of docfx.lib utilities such as "HrefUtility"

---
In reply to: [211816813](https://github.com/dotnet/docfx/pull/3253#discussion_r211816813) [](ancestors = 211816813)

[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/lib/attributes/ExternalFileAttribute.cs]: ok, let's figure that out later

---
In reply to: [211820033](https://github.com/dotnet/docfx/pull/3253#discussion_r211820033) [](ancestors = 211820033,211816813)

File: docs/specs/config.yml [hunk 5]
```diff
@@ -364,3 +364,13 @@ outputs:
...
   build.manifest:
 ---
+# Wrong locale
+os: windows
+inputs:
+  docfx.yml: |
+    locale: wrong/wrong
+outputs:
+  build.log: |
+    ["error","violate-schema","(Line: 0, Character: 0) Locale 'wrong/wrong' is not supported","docfx.yml"]
```
Comment (yufeih) [inline docs/specs/config.yml]: should this be a warning?

**Pull request merged**